### PR TITLE
Add browser API to enable page transitions plugins 

### PR DESCRIPTION
**Pull request by tannerlinsley**
This is very provisional, but I’m excited where it’s going :)

### CODE REVIEW

**Comment by tannerlinsley** [comment]
Here is a demo

```javascript
const React = require('react')
const { Transition } = require('react-move')

exports.componentRenderer = ({currentPage, previousPage}) => {
  console.log(currentPage, previousPage)
  return (
    <Transition
      data={[
        currentPage,
        previousPage
      ]}
      getKey={d => d.location.pathname}
      enter={d => ({
        transform: `translateX(100%)`,
        opacity: 1,
        zIndex: 1,
        boxShadow: '0 0 20px 0 rgba(0,0,0,0)',
      })}
      update={d => ({
        transform: `translateX(0%)`,
        opacity: 1,
        zIndex: 1,
        boxShadow: '0 0 100px 0 rgba(0,0,0,.5)'
      })}
      leave={d => ({
        transform: `translateX(-33%)`,
        opacity: .8,
        zIndex: 0
      })}
      ignore={['zIndex']}
      duration={700}
    >
      {nodes => (
        <div>
          {nodes.map(node => {
            const { component: Component, ...rest } = node.data
            return <div key={node.key} style={{
              position: 'absolute',
              left: 0,
              right: 0,
              top: 0,
              bottom: 0,
              background: 'white',
              ...node.state
            }}>
              <Component {...rest} />
            </div>
          })}
        </div>
      )}
    </Transition>
  )
}
```

**Comment by KyleAMathews** [comment]
@0x80

File: packages/gatsby/src/cache-dir/component-renderer.js [hunk 1]
```diff
@@ -55,22 +57,40 @@ class ComponentRenderer extends React.Component {
...
   }
 
   render() {
     if (this.state.pageResources) {
-      return createElement(this.state.pageResources.component, {
-        ...this.props,
-        ...this.state.pageResources.json,
+      const pluginResponses = apiRunner('componentRenderer', {
+        previousPage: {
+          ...this.props,
```
Comment (KyleAMathews) [inline packages/gatsby/src/cache-dir/component-renderer.js]: This should be `this.previousProps`

File: packages/gatsby/src/cache-dir/component-renderer.js [hunk 2]
```diff
@@ -55,22 +57,40 @@ class ComponentRenderer extends React.Component {
...
     }
     return false
   }
 
   render() {
     if (this.state.pageResources) {
-      return createElement(this.state.pageResources.component, {
-        ...this.props,
-        ...this.state.pageResources.json,
+      const pluginResponses = apiRunner('componentRenderer', {
```
Comment (KyleAMathews) [inline packages/gatsby/src/cache-dir/component-renderer.js]: Let's name this API `replacePageComponentRenderer` per https://www.gatsbyjs.org/docs/api-specification/#operators

**Comment by tannerlinsley** [comment]
I'll update these changes asap

**Comment by 0x80** [comment]
When I tried the example code, besides the mentioned previousPage workaround, the props.data of the rendered page seems to be undefined. So the result of the graphql query appears to be missing. I can't quite figure out where it went wrong.

**Comment by KyleAMathews** [comment]
@tannerlinsley hey I'm going to take a crack at finishing this — a client needs this functionality added ASAP.

**Comment by KyleAMathews** [comment]
Deploy preview failed.

Built with commit fdf1c51c005567d03dc1c14d131005078c805de4

https://app.netlify.com/sites/using-glamor/deploys/5963ed5ecf321c63a9d70aaf

**Comment by KyleAMathews** [comment]
Ok so didn't end up finishing this as the client and I decided to go with a simple entry animation for components not an exit & entry animation.

I did push a commit fixing up some things. Want to push this across the finish line still @tannerlinsley ?

**Comment by theiliad** [comment]
Is there an update on this? Any sample code?

**Comment by hochem** [comment]
@theiliad I added an example that doesn't rely on gatsby package modification in #1854

**Comment by LeKoArts** [comment]
Any news on this? 👍

**Comment by KyleAMathews** [comment]
Closing this as @stevensurgnier added an API with alternative approach to solving this issue https://github.com/gatsbyjs/gatsby/pull/2557

Thanks to everyone who's participated here! Check out the new API and let us know what you think!

**Pull request closed without merging**